Fix installation on Apple silicon Macs

hl-tam opened the pull request:
### What does this PR do?

This PR addresses two issues that blocks Salt from being installed on Apple silicon and a third issue that causes `salt` to crash on launch.

The first two issues are fixed by updating the version pins for `pyobjc` and `cffi`.  The crash is fixed by having Salt use the `HOMEBREW_PREFIX` environment variable instead of assuming that `brew` always installs to `/usr/local` (the default on Apple silicon Macs is `/opt/homebrew` and it's configurable on all platforms).

### What issues does this PR fix or reference?
Fixes: #59808 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

I don't think docs or tests are applicable to this PR. I'm not sure if this counts as a notable change for the change log.

### Commits signed with GPG?
No

hl-tam committed:
Update pyobjc to 7.1

Salt fails to install on macOS Big Sur because of this `pyobjc` issue:
    https://github.com/ronaldoussoren/pyobjc/issues/333

hl-tam committed:
Check $HOMEBREW_PREFIX when linking against libcrypto

When loading `libcrypto`, Salt checks for a Homebrew installation of `openssl`
at Homebrew's default prefix of `/usr/local`. However, on Apple Silicon Macs,
Homebrew's default installation prefix is `/opt/homebrew`. On all platforms,
the prefix is configurable.  If Salt doesn't find one of those `libcrypto`s,
it will fall back on the un-versioned `/usr/lib/libcrypto.dylib`, which will
cause the following crash:

    Application Specific Information:
    /usr/lib/libcrypto.dylib
    abort() called
    Invalid dylib load. Clients should not load the unversioned libcrypto dylib as it does not have a stable ABI.

This commit checks $HOMEBREW_PREFIX instead of hard-coding `/usr/local`.

hl-tam committed:
Update minimum required version of cffi on Darwin

On an Apple M1 Mac, `cffi` (required by `cryptography`) won't install due to

    https://foss.heptapod.net/pypy/cffi/-/issues/479

This is fixed in 1.14.4. A similar change was made to Windows dependencies in

    https://github.com/saltstack/salt/pull/59479

This change was tested on an M1 Mac running macOS 11.2.1 against both the
built-in Python 3.8.2 as well as Python 3.9.2 from Homebrew.

Ch3LL commented on salt/utils/rsax931.py:
```diff
@@ -32,8 +32,13 @@ def _find_libcrypto():
         # Find library symlinks in Homebrew locations.
-        lib = lib or glob.glob("/usr/local/opt/openssl/lib/libcrypto.dylib")
-        lib = lib or glob.glob("/usr/local/opt/openssl@*/lib/libcrypto.dylib")
+        brew_prefix = os.getenv("HOMEBREW_PREFIX", "/usr/local")
```
This will require test coverage and a changelog entry.

hl-tam committed:
Add changelog entry

hl-tam committed:
Make _find_libcrypto fail on Big Sur if it can't find a library

Right now, if `_find_libcrypto` can't find any externally-managed versions of
libcrypto, it will fall back on the pre-Catalina un-versioned system libcrypto.
This does not exist on Big Sur and it would be better to raise an exception
here rather than crashing later when trying to open it.

hl-tam committed:
Update _find_libcrypto tests

This commit simplifies the unit tests for _find_libcrypto by mocking out the
host's filesystem and testing the common libcrypto installations (brew, ports,
etc.) on Big Sur. It simplifies the tests for falling back on system versions
of libcrypto on previous versions of macOS.

krionbsd committed:
Merge branch 'master' into bigsur-arm-install

krionbsd committed:
Add test case

waynew commented on tests/unit/utils/test_rsax931.py:
```diff
@@ -218,6 +214,28 @@ def test_find_libcrypto_darwin_catalina(self):
+            if fnmatch.fnmatch(lib_path, i):
+                passed = True
+                break
+        self.assertTrue(passed)
```
**suggestion**: this test looks like it's simply re-implementing the code it's supposed to be testing. That's not very meaningful.

It would be more meaningful to do something like:

```
def test_when_homebrew_prefix_is_set_on_mac_lib_crypto_should_use_that_path():
    expected_lib_path = "/some/path/to/opt/openssl/lib/libcrypto.dylib"
    # patch getenv to return '/some/path/to'. Also probably need to patch glob
    lib_path = _find_libcrypto()
    assert lib_path == expected_lib_path
```

I would write a couple of tests - at *minimum* to cover the two scenarios that add `BREW_PREFIX`.

It's also preferred to add tests to `tests/pytests/`. If you need help, you can attend one of the [Test Clinics on Twitch](https://www.twitch.tv/saltprojectoss/schedule) and we'll help you get pointed in the right direction.

Ch3LL commented:
@hl-tam I ported some of the changes from this PR to here: https://github.com/saltstack/salt/pull/59828 so it can be included in the Aluminium release. We are hesitant in including the updated versions of the libraries in Aluminium since we are so close to releasing and this has caused us issues in the past when we update libraries right before a release without giving it enough time for testing. We will keep your changes to update the libraries here in master so it can be included in the next release Silicone, but will not make it into Aluminium. Thanks for the PR!

twangboy commented:
Has anyone tried actually building the packages with these changes....? Sometimes the dependencies can be a little tricky.

hl-tam commented:
I've added change log entry and rewrote the tests for `_find_libcrypto` on different macOS versions to run without relying on the underlying host filesystem. I didn't move them to `pytests` since all the other tests for the function are in this file and I don't know enough about the other OS code paths to feel comfortable porting the whole `test_rsax931.py` file over to the new test architecture.

Ch3LL commented on tests/unit/utils/test_rsax931.py:
```diff
@@ -174,49 +170,86 @@ def test_find_libcrypto_aix(self):
-    def test_find_libcrypto_with_system_and_not_catalina(self):
+    def test_find_libcrypto_with_system_before_catalina(self):
         """
         Test _find_libcrypto on a Catalina-like macOS host, simulate
```
Can you update the comment to reflect this tests before catalina and not on a catalina macos host.

Ch3LL reviewed (changes requested):
just a minor request

Ch3LL commented:
Also it looks like the tests are failing on windows.

hl-tam committed:
Fix description of test_find_libcrypto_with_system_before_catalina

hl-tam commented:
> Also it looks like the tests are failing on windows.
Ah, I see the issue: `sys.platform` isn't mocked.

Ch3LL committed:
Patch sys.platform for test_rsax931 tests

Ch3LL commented:
I fixed those tests on the freeze branch PR here: https://github.com/saltstack/salt/pull/59828/commits/0dd310d1708c1d97e4111c487df797ec18558aa1 

I can push that unless you had an idea for a different fix?

hl-tam commented:
> I can push that unless you had an idea for a different fix?

That's exactly what I was thinking, just didn't get around to it.  Thanks!

Ch3LL commented:
@twangboy are you good with this PR?

hl-tam committed:
Merge branch 'master' into bigsur-arm-install

Ch3LL commented:
ping @hl-tam looks like there is a merge conflict, most likely caused by this https://github.com/saltstack/salt/pull/60026

I'm guessing the only remaining change would be the cffi upgrade on your end?

hl-tam commented:
@Ch3LL sorry I missed this. With pyobjc upgraded by another PR, that still leaves CFFI as well as the Homebrew environment changes for Apple Silicon.

hl-tam commented:
Oh, never mind, I see that the Homebrew changes and associated tests have already been upstreamed. Installation works for me without issue on top of tree.  Thanks for grabbing all of this!
